### ISSUE TITLE
Automatic installation of a direct dependency's peer dependencies doesn't yield the expected result

### ISSUE THREAD

**Issue by alcuadrado** [pnpm/pnpm#8463]
### Verify latest release

- [X] I verified that the issue exists in the latest pnpm release

### pnpm version

9.8.0

### Which area(s) of pnpm are affected? (leave empty if unsure)

Dependencies resolver

### Reproduction steps

This script reproduces the behavior:

```sh
mkdir repro
cd repro
pnpm init
pnpm i react-dom
node -e "require('react')" 
```

### Describe the Bug

pnpm auto-installs peer dependencies, ensuring consistent versioning and avoiding missing peers.

It's auto-install resolution logic leads to a confusing scenario when you directly install a dependency that has peer dependencies, like `react-dom` in the script above. 

The peer dependency is installed and accessible to your dependency but not to your own package, giving the impression that it wasn't autoinstalled.

This behavior gets worse when you have transitive peer dependencies, creating more confusion.

### Expected Behavior

I believe your direct dependencies' peer dependencies should be available to your own project (i.e., in the main `node_modules`), and this is npm's behavior.

### Which Node.js version are you using?

22.3.0

### Which operating systems have you used?

- [ ] macOS
- [ ] Windows
- [X] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

A debian based official devcontainer

**Comment by zkochan** [comment]
This works correctly. If you want to use react in your code, you need to add it as a dependency.

**Comment by alcuadrado** [comment]
Thanks for the quick reply!

**Comment by alcuadrado** [comment]
@zkochan Thanks again for your answer, and sorry for getting back to this issue, but I maintain a project with a complex ecosystem of plugins, so peer dependencies are super important to me.

Your answer changed my understanding about peer dependencies and pnpm, but I have a few new questions now:

If I have these packages:

* `p1`: with peer dependencies `p2`, `p3`, and `p4`.
* `p2`: without peer dependencies.
* `p3`: with peer dependency `p4`.
* `p4`: without peer dependencies.

All their versions and version specs are compatible, and in my own files I only import `p1` and `p4`. 

1. Would installing `p1` and `p4` be enough? Will `pnpm` autoinstall `p2` and `p3`?
2. Will it install a single version of each of them?

In other words, should I instruct my users to only install the packages that they'll import?

**Comment by zkochan** [comment]
> Would installing p1 and p4 be enough? Will pnpm autoinstall p2 and p3?

Yes, that will be enough.

> Will it install a single version of each of them?

There are a few scenarios:

1. if these peer dependencies are in the peer dependencies of multiple packages and the ranges overlap, then pnpm will install a single version for each of them. So, if p2 is specified as `>=2.0.0` in one package and as `^2.0.0` in another one, then pnpm will install `p2@^2.0.0`.
2. if there is not overlap, pnpm will not install the peer dependency and it will print a warning.
3. if any of the peer dependencies is also installed as a regular non-direct dependency, then it really gets complicated and there is no guarantee that the peers will be singletons. Especially in your complex scenario, where peers have peers of their own.

**Comment by alcuadrado** [comment]
Thanks for your answer, @zkochan, I **really** appreciate it.

I think under (3) pnpm should have the freedom to install multiple versions.

In our plugin system we recommend these to plugin authors:
* If your plugin (P1) depends on another plugin (P2), that should be a peer dependency
* If P2 has peer dependencies, include them as your (P1's) peer dependencies too.

I think this forces everyone into the first situation.
